### PR TITLE
[Feature-6471]Cache Process definition in master

### LINKED ISSUE #6471: [Feature][Master] Cache Process definition in master
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

in order to reduce the db query times, cache process definitions in master.
procesDefinitionCacheMaps: {key: code-version, value: processDefintion}

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by lenboo**
close #6471  cache process definition in master

### CODE REVIEW

File: sql/dolphinscheduler_mysql.sql [hunk 1]
```diff
@@ -317,22 +317,24 @@ CREATE TABLE `t_ds_alertgroup`(
...
-  `command_param` text COMMENT 'json command parameters',
-  `task_depend_type` tinyint(4) DEFAULT NULL COMMENT 'Node dependency type: 0 current node, 1 forward, 2 backward',
-  `failure_strategy` tinyint(4) DEFAULT '0' COMMENT 'Failed policy: 0 end, 1 continue',
-  `warning_type` tinyint(4) DEFAULT '0' COMMENT 'Alarm type: 0 is not sent, 1 process is sent successfully, 2 process is sent failed, 3 process is sent successfully and all failures are sent',
-  `warning_group_id` int(11) DEFAULT NULL COMMENT 'warning group',
-  `schedule_time` datetime DEFAULT NULL COMMENT 'schedule time',
-  `start_time` datetime DEFAULT NULL COMMENT 'start time',
-  `executor_id` int(11) DEFAULT NULL COMMENT 'executor id',
-  `update_time` datetime DEFAULT NULL COMMENT 'update time',
+  `id`                        int(11)    NOT NULL AUTO_INCREMENT COMMENT 'key',
```
Comment (zhongjiajie) [inline sql/dolphinscheduler_mysql.sql:320]: I seem is unrelated change, just format modify, so as postgres.sql file

**Review by zhongjiajie — COMMENTED** [review]
Add some of minor suggestion

File: sql/dolphinscheduler_h2.sql [hunk 2]
```diff
@@ -328,6 +328,8 @@ CREATE TABLE t_ds_command
     worker_group              varchar(64),
     environment_code          bigint(20) DEFAULT '-1',
     dry_run                   int NULL DEFAULT 0,
+    process_instance_id       int(11) DEFAULT 0,
+    process_defintion_version int(11) DEFAULT 0,
```
Comment (zhongjiajie) [inline sql/dolphinscheduler_h2.sql]: I found out we have typo here, we use `process_definition_version` in `Command.java`, but all database field use `process_defintion_version`

```suggestion
    process_defintion_version int(11) DEFAULT 0,
    process_definition_version int(11) DEFAULT 0,
```

**Review by zhongjiajie — COMMENTED** [review]
Seem not CI failed, but we have typo in our database filed

File: sql/dolphinscheduler_h2.sql (hunk 2 above)
Comment (lenboo) [reply sql/dolphinscheduler_h2.sql]: good suggestion. thx.

File: sql/dolphinscheduler_mysql.sql (hunk 1 above)
Comment (lenboo) [reply sql/dolphinscheduler_mysql.sql:320]: yes, i add two fields in the end
and find the sql format is invalid, so i change the format at the same.

File: dolphinscheduler-server/src/main/resources/master.properties [hunk 3]
```diff
@@ -39,6 +39,9 @@
 # master commit task interval, the unit is millisecond
 #master.task.commit.interval=1000
 
+# master cache process definition
+#master.cache.process.definition=true
```
Comment (zhongjiajie) [inline dolphinscheduler-server/src/main/resources/master.properties:43]: Should we turn on cache by default
```suggestion
master.cache.process.definition=true
```

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/runner/MasterSchedulerService.java [hunk 4]
```diff
@@ -165,15 +181,19 @@ public void run() {
...
         Command command = findOneCommand();
         if (command != null) {
             logger.info("find one command: id: {}, type: {}", command.getId(), command.getCommandType());
             try {
                 ProcessInstance processInstance = processService.handleCommand(logger,
                         getLocalAddress(),
-                        this.masterConfig.getMasterExecThreads() - activeCount, command);
+                        command,
+                        processDefinitionCacheMaps);
+                if (!masterConfig.getMasterCacheProcessDefinition()
```
Comment (caishunfeng) [inline dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/runner/MasterSchedulerService.java:193]: Is it better to manage processDefinition cache by a unified manager?

File: dolphinscheduler-server/src/main/resources/master.properties (hunk 3 above)
Comment (lenboo) [reply dolphinscheduler-server/src/main/resources/master.properties:43]: we have already set the default value in the class MasterConfig.

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessService.java [hunk 5]
```diff
@@ -744,90 +735,74 @@ private Boolean checkCmdParam(Command command, Map<String, String> cmdParam) {
...
      */
-    private ProcessInstance constructProcessInstance(Command command, String host) {
+    private ProcessInstance constructProcessInstance(Command command, String host, HashMap<String, ProcessDefinition> processDefinitionCacheMaps) {
         ProcessInstance processInstance;
+        ProcessDefinition processDefinition;
         CommandType commandType = command.getCommandType();
-        Map<String, String> cmdParam = JSONUtils.toMap(command.getCommandParam());
-
-        ProcessDefinition processDefinition = getProcessDefinitionByCommand(command.getProcessDefinitionCode(), cmdParam);
+        String key = String.format("%d-%d", command.getProcessDefinitionCode(), command.getProcessDefinitionVersion());
```
Comment (caishunfeng) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessService.java:747]: Any other else? Or it just use for constructProcessInstance?

Comment (lenboo) [reply dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessService.java:747]: - Is it better to manage processDefinition cache by a unified manager?
yes, you are right, we can use a unified cache manager. we need to optimize this caching method next.

File: dolphinscheduler-server/src/main/resources/master.properties (hunk 3 above)
Comment (zhongjiajie) [reply dolphinscheduler-server/src/main/resources/master.properties:43]: For this situation, I think we better add comment in L42 to desc `# master cache process definition, default true` make it more clearer

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessService.java (hunk 5 above)
Comment (lenboo) [reply dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessService.java:747]: > Any other else? Or it just use for constructProcessInstance?

maybe the task definitions can also be cached
and a clean-up mechanism can also be added to the cache queue, eg: clean cache queue once a day?

File: dolphinscheduler-server/src/main/resources/master.properties (hunk 3 above)
Comment (lenboo) [reply dolphinscheduler-server/src/main/resources/master.properties:43]: you are right.
thank you for your suggestion.

**Review by CalvinKirs — APPROVED** [review]

**Review by zhongjiajie — APPROVED** [review]
LGTM now

**Review by zhuangchong — APPROVED** [review]
